controllers/cluster_controller.go

```diff
@@ -63,20 +69,74 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, err
 	}
 
-	// Return early if the object or Cluster is paused.
+	// Return if the Cluster is paused.
 	if annotations.IsPaused(cluster, cluster) {
+		r.Log.Info("The cluster is paused.")
 		return ctrl.Result{}, nil
 	}
 
-	// your logic here
+	// Return if the Cluster is deleted.
+	if !cluster.DeletionTimestamp.IsZero() {
+		r.Log.Info("The cluster is deleted.")
+		return ctrl.Result{}, nil
+	}
+
+	// Return if there is no upgrade time scheduled.
+	if getClusterUpgradeTimeAnnotation(cluster) == "" {
+		r.Log.Info("The cluster has no upgrade scheduled.")
+		return ctrl.Result{}, nil
+	}
+
+	// Return if the upgrade release version is not specified.
+	if getClusterUpgradeVersionAnnotation(cluster) == "" {
+		r.Log.Info("The scheduled update at %v can not proceed because no target release version has been set via annotation %v.", getClusterUpgradeTimeAnnotation(cluster), annotation.AWSUpdateScheduleTargetRelease)
+		return ctrl.Result{}, nil
+	}
+
+	return r.ReconcileUpgrade(ctx, cluster)
+}
+
+func (r *ClusterReconciler) ReconcileUpgrade(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	upgradeTime, err := time.Parse(time.RFC822, getClusterUpgradeTimeAnnotation(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return if the scheduled upgrade time is not reached yet.
+	if !upgradeTimeReached(upgradeTime) {
+		r.Log.Info(fmt.Sprintf("The scheduled update time is not reached yet. Cluster will be upgraded at %v.", upgradeTime))
+		return ctrl.Result{}, nil
+	}
+
+	currentVersion, err := semver.New(getClusterReleaseVersionLabel(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	targetVersion, err := semver.New(getClusterUpgradeVersionAnnotation(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
I would try print some log error here to better understand why some update was not applied

might be also a good idea to have a validation webhook on the annotations so somebody does not apply something wrongly and miss update window because if ot

Reply:
I have added the printing of errors. Let's add validation in the admission controller.